File: manage_employee_schedule.py
from typing import List, Dict, Set
from enum import Enum
import logging
import yaml
import random
logging.basicConfig(level=logging.INFO)

class Shift(Enum):
    MORNING = 0
    AFTERNOON = 1
    EVENING = 2

class Day(Enum):
    MONDAY = 1
    TUESDAY = 2
    WEDNESDAY = 3
    THURSDAY = 4
    FRIDAY = 5
    SATURDAY = 6
    SUNDAY = 7

class ManageSchedule:
    def __init__(self):
        self.employeeSchedule = {} # {name: [(day1,shift1), (day2,shift2), ...]}
        self.employeeNumberPerShiftDay = {} # {day: [[morning,x],[afternoon,y],[evening,z]]}
        self.preferences = {} # {name: [(day1,shift1), (day2,shift2), ...]}

        self.initializeEmployeePerShiftDay()
        
    def getPreference(self, filename: str) -> None:
        '''
        set a preference of employees' schedule
        return true if successful
        '''
        with open(filename, 'r') as file:
            preferences = yaml.safe_load(file)
        logging.debug(f'finish loading preferences: {preferences}')

        for i, employee in enumerate(preferences['employees']):
            slots = []
            for j, slot in enumerate(employee['preferences']):
                day = slot['day']
                shift = slot['time']
                slots.append((day, shift))
            self.preferences[employee['name'].upper()] = slots

    def writeOutput(self, filename: str) -> None:
        '''
        write the output to the file
        return true if successful
        '''
        with open(filename, 'w') as file:
            yaml.dump(self.employeeSchedule, file)
        logging.debug(f'successfully wrote the schedule to {filename}')
    
    def initializeEmployeePerShiftDay(self) -> None:
        '''
        initialize employeePerShiftDay with empty list
        '''
        for day in (Day):
            self.employeeNumberPerShiftDay[day.name] = [0, 0, 0]
        logging.info('finish initializing employeeNumberPerShiftDay')
        
    def assignShift(self, maxWorkDay: int = 5) -> None:
        '''
        assign shift to employees

        This method will assign shift to employees based on their preferences.
        Use this following methods
        '''
        for name, _ in self.preferences.items():
            if not self.isOneShiftPerDay(name):
                logging.error(f'employee {name} has more than one preference shift per day!')
                break
            if self.getNumofWorkPreference(name) > maxWorkDay:
                logging.error(f'employee {name} has preference work more than 5 days!')
                break

        for name, preference in self.preferences.items():
            for day, shift in preference:
                if shift == 'morning':
                    numMorning, _, _ = self.getEmployeeNumPerShiftDay(day)
                    if self.getNumberofWorkAssigned(name) < maxWorkDay and numMorning < 2:
                        self.updateEmployeeNumPerShiftDay(day, shift)
                        self.setEmployeeSchedule(name, day, shift)
                    else:
                        # assign alternate shift
                        shift = self.findAvailableShiftDay(name, day, maxWorkDay)
                        if shift is not None:
                            self.setEmployeeSchedule(name, day, shift)
                            self.updateEmployeeNumPerShiftDay(day, shift)
                        else:
                            # assign shift to the employee next available day
                            for next_day in self.findNextDays(day):
                                shift = self.findAvailableShiftDay(name, next_day, maxWorkDay)
                                if shift is None:
                                    continue
                                else:    
                                    self.setEmployeeSchedule(name, next_day, shift)
                                    self.updateEmployeeNumPerShiftDay(next_day, shift)
                                    break
                elif shift == 'afternoon':
                    _, numAfternoon, _ = self.getEmployeeNumPerShiftDay(day)
                    if self.getNumberofWorkAssigned(name) < maxWorkDay and numAfternoon < 2:
                        self.updateEmployeeNumPerShiftDay(day, shift)
                        self.setEmployeeSchedule(name, day, shift)
                    else:
                        # assign alternate shift
                        shift = self.findAvailableShiftDay(name, day, maxWorkDay)
                        if shift is not None:
                            self.setEmployeeSchedule(name, day, shift)
                            self.updateEmployeeNumPerShiftDay(day, shift)
                        else:
                            # assign shift to the employee next available day
                            for next_day in self.findNextDays(day):
                                shift = self.findAvailableShiftDay(name, next_day, maxWorkDay)
                                if shift is None:
                                    continue
                                else:    
                                    self.setEmployeeSchedule(name, next_day, shift)
                                    self.updateEmployeeNumPerShiftDay(next_day, shift)
                                    break
                elif shift == 'evening':
                    _, _, numEvening = self.getEmployeeNumPerShiftDay(day)
                    if self.getNumberofWorkAssigned(name) < maxWorkDay and numEvening < 2:
                        self.updateEmployeeNumPerShiftDay(day, shift)
                        self.setEmployeeSchedule(name, day, shift)
                    else:
                        # assign alternate shift
                        shift = self.findAvailableShiftDay(name, day, maxWorkDay)
                        if shift is not None:
                            self.setEmployeeSchedule(name, day, shift)
                            self.updateEmployeeNumPerShiftDay(day, shift)
                        else:
                            # assign shift to the employee next available day
                            for next_day in self.findNextDays(day):
                                shift = self.findAvailableShiftDay(name, next_day, maxWorkDay)
                                if shift is None:
                                    continue
                                else:    
                                    self.setEmployeeSchedule(name, next_day, shift)
                                    self.updateEmployeeNumPerShiftDay(next_day, shift)
                                    break
                else:
                    logging.error(f'invalid shift: {shift}')
                    ValueError(f'invalid shift: {shift}')
                    break

        # fill under staffed shifts
        self.fillUnderStaffedÍhifts()       
                        
    def isOneShiftPerDay(self, name: str) -> bool:
        '''
        return true if one shift per day of the employee

        isOneShiftPerDay: this condition will go through preferences and check if the employee has only one shift per day
        e.g. compare tuple (day1,shift1) and (day2, shift2) if day1 == day2 then isOneShiftPerday is false

        datastructure: preferences
        is used in assignShift method
        '''
        perference_list = self.preferences[name.upper()]
        for i in range(len(perference_list)):
            for j in range(i+1, len(perference_list)):
                if perference_list[i][0].upper() == perference_list[j][0].upper():
                    return False
        return True
    
    def getNumofWorkPreference(self, name: str) -> int:
        '''
        return the number of days the employee works

        e.g. for employee A, count the number of tuples (day,shift) in preferences.
        
        datastructure: preferences
        is used in assignShift method
        '''
        return len(self.preferences[name.upper()])
    
    def getNumberofWorkAssigned(self, name: str) -> int:
        '''
        return the number of days the employee works

        e.g. for employee A, count the number of tuples (day,shift) in employeeSchedule.
        
        datastructure: employeeSchedule
        is used in assignShift method
        '''
        return len(self.employeeSchedule.get(name.upper(), []))

    def getEmployeeNumPerShiftDay(self, day: str) -> tuple:
        '''
        return the number of employees per shift of the day (morning, afternoon, evening)
        in the following format (shift, number of employees)
        
        e.g. 5, 3, 4 # morning, afternoon, evening

        datastructure: employeeNumberPerShiftDay
        is used in assignShift method

        '''
        match day.upper():
            case 'MON':
                numMorning = self.employeeNumberPerShiftDay[Day.MONDAY.name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day.MONDAY.name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day.MONDAY.name][Shift.EVENING.value]
            case 'TUE':
                numMorning = self.employeeNumberPerShiftDay[Day.TUESDAY.name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day.TUESDAY.name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day.TUESDAY.name][Shift.EVENING.value]
            case 'WED':
                numMorning = self.employeeNumberPerShiftDay[Day.WEDNESDAY.name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day.WEDNESDAY.name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day.WEDNESDAY.name][Shift.EVENING.value]
            case 'THU':
                numMorning = self.employeeNumberPerShiftDay[Day.THURSDAY.name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day.THURSDAY.name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day.THURSDAY.name][Shift.EVENING.value]
            case 'FRI':
                numMorning = self.employeeNumberPerShiftDay[Day.FRIDAY.name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day.FRIDAY.name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day.FRIDAY.name][Shift.EVENING.value]
            case 'SAT':
                numMorning = self.employeeNumberPerShiftDay[Day.SATURDAY.name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day.SATURDAY.name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day.SATURDAY.name][Shift.EVENING.value]
            case 'SUN':
                numMorning = self.employeeNumberPerShiftDay[Day.SUNDAY.name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day.SUNDAY.name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day.SUNDAY.name][Shift.EVENING.value]
            case _:
                numMorning = self.employeeNumberPerShiftDay[Day[day.upper()].name][Shift.MORNING.value]
                numAfternoon = self.employeeNumberPerShiftDay[Day[day.upper()].name][Shift.AFTERNOON.value]
                numEvening = self.employeeNumberPerShiftDay[Day[day.upper()].name][Shift.EVENING.value]
        
        return (numMorning, numAfternoon, numEvening)
    
    def updateEmployeeNumPerShiftDay(self, day: str, shift: str) -> None:
        '''
        update the number of employees per shift of the day (morning, afternoon, evening)
        '''
        match day.upper():
            case 'MON':
                self.employeeNumberPerShiftDay[Day.MONDAY.name][Shift[shift.upper()].value] += 1
            case 'TUE':
                self.employeeNumberPerShiftDay[Day.TUESDAY.name][Shift[shift.upper()].value] += 1
            case 'WED':
                self.employeeNumberPerShiftDay[Day.WEDNESDAY.name][Shift[shift.upper()].value] += 1
            case 'THU':
                self.employeeNumberPerShiftDay[Day.THURSDAY.name][Shift[shift.upper()].value] += 1
            case 'FRI':
                self.employeeNumberPerShiftDay[Day.FRIDAY.name][Shift[shift.upper()].value] += 1
            case 'SAT':
                self.employeeNumberPerShiftDay[Day.SATURDAY.name][Shift[shift.upper()].value] += 1
            case 'SUN':
                self.employeeNumberPerShiftDay[Day.SUNDAY.name][Shift[shift.upper()].value] += 1
            case _:
                self.employeeNumberPerShiftDay[Day[day.upper()].name][Shift[shift.upper()].value] += 1
        return None
    
    def setEmployeeSchedule(self, name: str, day: str, shift: str) -> None:
        '''
        set the employee schedule
        '''
        match day.upper():
                case 'MON':
                    schedule = (Day.MONDAY.name, Shift[shift.upper()].name)
                case 'TUE':
                    schedule = (Day.TUESDAY.name, Shift[shift.upper()].name)
                case 'WED':
                    schedule = (Day.WEDNESDAY.name, Shift[shift.upper()].name)
                case 'THU':
                    schedule = (Day.THURSDAY.name, Shift[shift.upper()].name)
                case 'FRI':
                    schedule = (Day.FRIDAY.name, Shift[shift.upper()].name)
                case 'SAT':
                    schedule = (Day.SATURDAY.name, Shift[shift.upper()].name)
                case 'SUN':
                    schedule = (Day.SUNDAY.name, Shift[shift.upper()].name)
                case _:
                    schedule = (Day[day.upper()].name, Shift[shift.upper()].name)
        
        if name.upper() not in self.employeeSchedule.keys():
            self.employeeSchedule[name.upper()] = []
            self.employeeSchedule[name.upper()].append(schedule)
        else:
            self.employeeSchedule[name.upper()].append(schedule)

        return None
    
    def findAvailableShiftDay(self, name: str, day: str, maxWorkDay: int) -> str:
        '''
        find available shift
        '''
        numMorning, numAfternoon, numEvening = self.getEmployeeNumPerShiftDay(day)
        for shift in ['morning', 'afternoon', 'evening']:
            if shift == 'morning' and numMorning < 2:
                if self.getNumberofWorkAssigned(name) < maxWorkDay:
                    return 'morning'
            elif shift == 'afternoon' and numAfternoon < 2:
                if self.getNumberofWorkAssigned(name) < maxWorkDay:
                    return 'afternoon'
            elif shift == 'evening' and numEvening < 2:
                if self.getNumberofWorkAssigned(name) < maxWorkDay:
                    return 'evening'
            else:
                return None
        
    def findNextDays(self, day: str) -> tuple:
        '''
        return the next days
        '''
        match day.upper():
            case 'MON':
                return ('TUE', 'WED', 'THU', 'FRI', 'SAT', 'SUN')
            case 'TUE':
                return ('WED', 'THU', 'FRI', 'SAT', 'SUN', 'MON')
            case 'WED':
                return ('THU', 'FRI', 'SAT', 'SUN', 'MON', 'TUE')
            case 'THU':
                return ('FRI', 'SAT', 'SUN', 'MON', 'TUE', 'WED')
            case 'FRI':
                return ('SAT', 'SUN', 'MON', 'TUE', 'WED', 'THU')
            case 'SAT':
                return ('SUN', 'MON', 'TUE', 'WED', 'THU', 'FRI')
            case 'SUN':
                return ('MON', 'TUE', 'WED', 'THU', 'FRI', 'SAT')
            case _:
                return ()
    
    def fillUnderStaffedÍhifts(self):
        '''
        fill the under staffed shifts
        '''
        availableEmployees = [name for name in self.preferences.keys() if self.getNumberofWorkAssigned(name) < 5]
        for day, shifts in self.employeeNumberPerShiftDay.items():
            morning, afternoon, evening = self.getEmployeeNumPerShiftDay(day)
            while morning < 2 and availableEmployees:
                selectedName = random.choice(availableEmployees)
                self.setEmployeeSchedule(selectedName, day, 'morning')
                self.updateEmployeeNumPerShiftDay(day, 'morning')
                morning, _, _ = self.getEmployeeNumPerShiftDay(day)
                if self.getNumberofWorkAssigned(selectedName) == 5:
                    availableEmployees.remove(selectedName)
            while afternoon < 2 and availableEmployees:
                selectedName = random.choice(availableEmployees)
                self.setEmployeeSchedule(selectedName, day, 'afternoon')
                self.updateEmployeeNumPerShiftDay(day, 'afternoon')
                _, afternoon, _ = self.getEmployeeNumPerShiftDay(day)
                if self.getNumberofWorkAssigned(selectedName) == 5:
                    availableEmployees.remove(selectedName)
            while evening < 2 and availableEmployees:
                selectedName = random.choice(availableEmployees)
                self.setEmployeeSchedule(selectedName, day, 'evening')
                self.updateEmployeeNumPerShiftDay(day, 'evening')
                _, _, evening = self.getEmployeeNumPerShiftDay(day)
                if self.getNumberofWorkAssigned(selectedName) == 5:
                    availableEmployees.remove(selectedName)
    
        
if __name__ == '__main__':
    schedule = ManageSchedule()

    schedule.getPreference('input/preference_schedule.yaml')
    schedule.assignShift(maxWorkDay=5)
    logging.info(f'successfully assigned shift to employees')
    schedule.writeOutput('output/schedule.yaml')
    logging.info(f'successfully wrote the schedule to output/schedule.yaml')

File: manage_employee_schedule.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <optional>
#include <random>
#include <set>
#include <string>
#include <vector>
#include <yaml-cpp/yaml.h>


enum class Shift {
    MORNING,
    AFTERNOON,
    EVENING,
    NONE
};

enum class Day {
    MONDAY,
    TUESDAY,
    WEDNESDAY,
    THURSDAY,
    FRIDAY,
    SATURDAY,
    SUNDAY
};

class ManageSchedule {
public:
    ManageSchedule() : debug(true) {
        initializeEmployeePerShiftDay();
    }

    void getPreference(const std::string& filename) {
        YAML::Node preferences = YAML::LoadFile(filename);
        for (const auto& employee : preferences["employees"]) {
            if (debug)
            {
                std::cout << "Employee: " << employee["name"].as<std::string>() << std::endl;
            }
            std::vector<std::pair<Day, Shift>> slots;
            for (const auto& slot : employee["preferences"]) {
                Day day = stringToDay(slot["day"].as<std::string>());
                Shift shift = stringToShift(slot["time"].as<std::string>());
                slots.emplace_back(day, shift);
                // debug print day, shift
                if (debug)
                {
                    // to ensure that stringToDay and stringToShift are working correctly
                    std::cout << "Day: " << static_cast<int>(day) << " Shift: " << static_cast<int>(shift) << std::endl;
                }
                
            }
            preferences_[employee["name"].as<std::string>()] = slots;
        }
    }

    void writeOutput(const std::string& filename) {
        YAML::Emitter out;
        out << YAML::BeginMap;
        for (const auto& [name, schedule] : employeeSchedule_) {
            out << YAML::Key << name << YAML::Value << YAML::BeginSeq;
            for (const auto& [day, shift] : schedule) {
                out << YAML::BeginMap;
                out << YAML::Key << "day" << YAML::Value << dayToString(day);
                out << YAML::Key << "shift" << YAML::Value << shiftToString(shift);
                out << YAML::EndMap;
            }
            out << YAML::EndSeq;
        }
        out << YAML::EndMap;

        std::ofstream fout(filename);
        fout << out.c_str();
    }

    void assignShift(int maxWorkDay = 5) {
        for (const auto& [name, _] : preferences_) {
            if (!isOneShiftPerDay(name)) {
                std::cerr << "Employee " << name << " has more than one preference shift per day!" << std::endl;
                break;
            }
            if (getNumofWorkPreference(name) > maxWorkDay) {
                std::cerr << "Employee " << name << " has preference work more than 5 days!" << std::endl;
                break;
            }
        }

        for (const auto& [name, preference] : preferences_) {
            for (const auto& [day, shift] : preference) {
                if (shift == Shift::MORNING) {
                    auto [numMorning, unused1, unused2] = getEmployeeNumPerShiftDay(day);
                    if (getNumberofWorkAssigned(name) < maxWorkDay && numMorning < 2) {
                        updateEmployeeNumPerShiftDay(day, shift);
                        setEmployeeSchedule(name, day, shift);
                    } else {
                        auto alternateShift = findAvailableShiftDay(name, day, maxWorkDay);
                        if (alternateShift != Shift::NONE) {
                            setEmployeeSchedule(name, day, alternateShift);
                            updateEmployeeNumPerShiftDay(day, alternateShift);
                        } else {
                            auto next_days = findNextDays(day);
                            for (const auto& next_day : next_days) {
                                auto nextShift = findAvailableShiftDay(name, next_day, maxWorkDay);
                                if (nextShift == Shift::NONE) {
                                    continue;
                                } else {
                                    setEmployeeSchedule(name, next_day, nextShift);
                                    updateEmployeeNumPerShiftDay(next_day, nextShift);
                                    break;
                                }
                            }
                        }
                    }
                } else if (shift == Shift::AFTERNOON) {
                    auto [unused1, numAfternoon, unused2] = getEmployeeNumPerShiftDay(day);
                    if (getNumberofWorkAssigned(name) < maxWorkDay && numAfternoon < 2) {
                        updateEmployeeNumPerShiftDay(day, shift);
                        setEmployeeSchedule(name, day, shift);
                    } else {
                        auto alternateShift = findAvailableShiftDay(name, day, maxWorkDay);
                        if (alternateShift != Shift::NONE) {
                            setEmployeeSchedule(name, day, alternateShift);
                            updateEmployeeNumPerShiftDay(day, alternateShift);
                        } else {
                            auto next_days = findNextDays(day);
                            for (const auto& next_day : next_days) {
                                auto nextShift = findAvailableShiftDay(name, next_day, maxWorkDay);
                                if (nextShift == Shift::NONE) {
                                    continue;
                                } else {
                                    setEmployeeSchedule(name, next_day, nextShift);
                                    updateEmployeeNumPerShiftDay(next_day, nextShift);
                                    break;
                                }
                            }
                        }
                    }
                } else if (shift == Shift::EVENING) {
                    auto [unused1, unused2, numEvening] = getEmployeeNumPerShiftDay(day);
                    if (getNumberofWorkAssigned(name) < maxWorkDay && numEvening < 2) {
                        updateEmployeeNumPerShiftDay(day, shift);
                        setEmployeeSchedule(name, day, shift);
                    } else {
                        auto alternateShift = findAvailableShiftDay(name, day, maxWorkDay);
                        if (alternateShift != Shift::NONE) {
                            setEmployeeSchedule(name, day, alternateShift);
                            updateEmployeeNumPerShiftDay(day, alternateShift);
                        } else {
                            auto next_days = findNextDays(day);
                            for (const auto& next_day : next_days) {
                                auto nextShift = findAvailableShiftDay(name, next_day, maxWorkDay);
                                if (nextShift == Shift::NONE) {
                                    continue;
                                } else {
                                    setEmployeeSchedule(name, next_day, nextShift);
                                    updateEmployeeNumPerShiftDay(next_day, nextShift);
                                    break;
                                }
                            }
                        }
                    }
                } else {
                    std::cerr << "Invalid shift: " << static_cast<int>(shift) << std::endl;
                    throw std::invalid_argument("Invalid shift");
                }
            }
        }

        fillUnderStaffedShifts();
        printEmployeeSchedule();
    }

private:
    bool debug;
    std::map<std::string, std::vector<std::pair<Day, Shift>>> employeeSchedule_;
    std::map<std::string, std::vector<std::pair<Day, Shift>>> preferences_;
    std::map<Day, std::array<int, 3>> employeeNumberPerShiftDay_;

    void printEmployeeSchedule() {
        for (const auto& [name, schedule] : employeeSchedule_) {
            std::cout << name << ": ";
            for (const auto& [day, shift] : schedule) {
                std::cout << dayToString(day) << " " << shiftToString(shift) << ", ";
            }
            std::cout << std::endl;
        }
    }

    void initializeEmployeePerShiftDay() {
        for (int i = static_cast<int>(Day::MONDAY); i <= static_cast<int>(Day::SUNDAY); ++i) {
            employeeNumberPerShiftDay_[static_cast<Day>(i)] = {0, 0, 0};
        }
    }

    bool isOneShiftPerDay(const std::string& name) {
        const auto& preference_list = preferences_[name];
        for (size_t i = 0; i < preference_list.size(); ++i) {
            for (size_t j = i + 1; j < preference_list.size(); ++j) {
                if (preference_list[i].first == preference_list[j].first) {
                    return false;
                }
            }
        }
        return true;
    }

    int getNumofWorkPreference(const std::string& name) {
        return preferences_[name].size();
    }

    int getNumberofWorkAssigned(const std::string& name) {
        if (employeeSchedule_.find(name) != employeeSchedule_.end()) {
            return employeeSchedule_[name].size();
        }
        return 0;
    }

    std::tuple<int, int, int> getEmployeeNumPerShiftDay(Day day) {
        const auto& shifts = employeeNumberPerShiftDay_[day];
        return std::make_tuple(shifts[0], shifts[1], shifts[2]);
    }

    void updateEmployeeNumPerShiftDay(Day day, Shift shift) {
        if (shift == Shift::MORNING || shift == Shift::AFTERNOON || shift == Shift::EVENING) {
            employeeNumberPerShiftDay_[day][static_cast<int>(shift)]++;
        } else {
            throw std::invalid_argument("Invalid shift");
        }
    }

    void setEmployeeSchedule(const std::string& name, Day day, Shift shift) {
        employeeSchedule_[name].emplace_back(day, shift);
    }

    Shift findAvailableShiftDay(const std::string& name, Day day, int maxWorkDay) {
        if (getNumberofWorkAssigned(name) == maxWorkDay) {
            return Shift::NONE; // No available shift
        }
        auto [numMorning, numAfternoon, numEvening] = getEmployeeNumPerShiftDay(day);
        if (numMorning < 2) {
            return Shift::MORNING;
        } 
        else if (numAfternoon < 2) {
            return Shift::AFTERNOON;
        }
        else if (numEvening < 2) {
            return Shift::EVENING;
        } else {
            return Shift::NONE; // No available shift
        }
        
    }

    std::vector<Day> findNextDays(Day day) {
        switch (day) {
            case Day::MONDAY: return {Day::TUESDAY, Day::WEDNESDAY, Day::THURSDAY, Day::FRIDAY, Day::SATURDAY, Day::SUNDAY};
            case Day::TUESDAY: return {Day::WEDNESDAY, Day::THURSDAY, Day::FRIDAY, Day::SATURDAY, Day::SUNDAY, Day::MONDAY};
            case Day::WEDNESDAY: return {Day::THURSDAY, Day::FRIDAY, Day::SATURDAY, Day::SUNDAY, Day::MONDAY, Day::TUESDAY};
            case Day::THURSDAY: return {Day::FRIDAY, Day::SATURDAY, Day::SUNDAY, Day::MONDAY, Day::TUESDAY, Day::WEDNESDAY};
            case Day::FRIDAY: return {Day::SATURDAY, Day::SUNDAY, Day::MONDAY, Day::TUESDAY, Day::WEDNESDAY, Day::THURSDAY};
            case Day::SATURDAY: return {Day::SUNDAY, Day::MONDAY, Day::TUESDAY, Day::WEDNESDAY, Day::THURSDAY, Day::FRIDAY};
            case Day::SUNDAY: return {Day::MONDAY, Day::TUESDAY, Day::WEDNESDAY, Day::THURSDAY, Day::FRIDAY, Day::SATURDAY};
            default: return {};
        }
    }

    void fillUnderStaffedShifts() {
        std::vector<std::string> availableEmployees;
        for (const auto& [name, _] : preferences_) {
            if (getNumberofWorkAssigned(name) < 5) {
                availableEmployees.push_back(name);
            }
        }

        for (const auto& [day, shifts] : employeeNumberPerShiftDay_) {
            auto [morning, afternoon, evening] = getEmployeeNumPerShiftDay(day);
            while (morning < 2 && !availableEmployees.empty()) {
                std::string selectedName = availableEmployees[rand() % availableEmployees.size()];
                setEmployeeSchedule(selectedName, day, Shift::MORNING);
                updateEmployeeNumPerShiftDay(day, Shift::MORNING);
                morning = std::get<0>(getEmployeeNumPerShiftDay(day));
                if (getNumberofWorkAssigned(selectedName) == 5) {
                    availableEmployees.erase(std::remove(availableEmployees.begin(), availableEmployees.end(), selectedName), availableEmployees.end());
                }
            }
            while (afternoon < 2 && !availableEmployees.empty()) {
                std::string selectedName = availableEmployees[rand() % availableEmployees.size()];
                setEmployeeSchedule(selectedName, day, Shift::AFTERNOON);
                updateEmployeeNumPerShiftDay(day, Shift::AFTERNOON);
                afternoon = std::get<1>(getEmployeeNumPerShiftDay(day));
                if (getNumberofWorkAssigned(selectedName) == 5) {
                    availableEmployees.erase(std::remove(availableEmployees.begin(), availableEmployees.end(), selectedName), availableEmployees.end());
                }
            }
            while (evening < 2 && !availableEmployees.empty()) {
                std::string selectedName = availableEmployees[rand() % availableEmployees.size()];
                setEmployeeSchedule(selectedName, day, Shift::EVENING);
                updateEmployeeNumPerShiftDay(day, Shift::EVENING);
                evening = std::get<2>(getEmployeeNumPerShiftDay(day));
                if (getNumberofWorkAssigned(selectedName) == 5) {
                    availableEmployees.erase(std::remove(availableEmployees.begin(), availableEmployees.end(), selectedName), availableEmployees.end());
                }
            }
        }
    }

    Day stringToDay(const std::string& day) {
        if (day == "Mon") return Day::MONDAY;
        if (day == "Tue") return Day::TUESDAY;
        if (day == "Wed") return Day::WEDNESDAY;
        if (day == "Thu") return Day::THURSDAY;
        if (day == "Fri") return Day::FRIDAY;
        if (day == "Sat") return Day::SATURDAY;
        if (day == "Sun") return Day::SUNDAY;

        if (day == "Monday") return Day::MONDAY;
        if (day == "Tuesday") return Day::TUESDAY;
        if (day == "Wednesday") return Day::WEDNESDAY;
        if (day == "Thursday") return Day::THURSDAY;
        if (day == "Friday") return Day::FRIDAY;
        if (day == "Saturday") return Day::SATURDAY;
        if (day == "Sunday") return Day::SUNDAY;
        
        throw std::invalid_argument("Invalid day");
    }

    std::string dayToString(Day day) {
        switch (day) {
            case Day::MONDAY: return "MON";
            case Day::TUESDAY: return "TUE";
            case Day::WEDNESDAY: return "WED";
            case Day::THURSDAY: return "THU";
            case Day::FRIDAY: return "FRI";
            case Day::SATURDAY: return "SAT";
            case Day::SUNDAY: return "SUN";
            default: throw std::invalid_argument("Invalid day");
        }
    }

    Shift stringToShift(const std::string& shift) {
        if (shift == "morning") return Shift::MORNING;
        if (shift == "afternoon") return Shift::AFTERNOON;
        if (shift == "evening") return Shift::EVENING;
        throw std::invalid_argument("Invalid shift");
    }

    std::string shiftToString(Shift shift) {
        switch (shift) {
            case Shift::MORNING: return "morning";
            case Shift::AFTERNOON: return "afternoon";
            case Shift::EVENING: return "evening";
            default: throw std::invalid_argument("Invalid shift");
        }
    }
};

int main() {
    try {
        ManageSchedule schedule;

        schedule.getPreference("/Users/peeratienthong/manage_employee_schedule/input/preference_schedule.yaml");
        schedule.assignShift(5);
        std::cout << "Successfully assigned shift to employees" << std::endl;
        schedule.writeOutput("/Users/peeratienthong/manage_employee_schedule/output/schedule_cpp.yaml");
        std::cout << "Successfully wrote the schedule to output/schedule.yaml" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << e.what() << std::endl;
        return 1;
    }
    return 0;
}
